accounts/abi/bind/precompiletemplate.go

```diff
@@ -0,0 +1,390 @@
+	AllowList bool                   // Indicator whether the contract uses AllowList precompile
+	Funcs     map[string]*tmplMethod // Contract functions that include both Calls + Transacts in tmplContract
+}
+
+// tmplSourcePrecompileGo is the Go precompiled source template.
+const tmplSourcePrecompileGo = `
+// Code generated
+// This file is a generated precompile contract with stubbed abstract functions.
+// The file is generated by a template. Please inspect every code and comment in this file before use.
+
+// There are some must-be-done changes waiting in the file. Each area requiring you to add your code is marked with CUSTOM CODE to make them easy to find and modify.
+// Additionally there are other files you need to edit to activate your precompile.
+// These areas are highlighted with comments "ADD YOUR PRECOMPILE HERE".
+// For testing take a look at other precompile tests in core/stateful_precompile_test.go
+
+/* General guidelines for precompile development:
+1- Read the comment and set a suitable contract address in precompile/params.go. E.g:
+	{{.Contract.Type}}Address = common.HexToAddress("ASUITABLEHEXADDRESS")
+2- Set gas costs here
+3- It is recommended to only modify code in the highlighted areas marked with "CUSTOM CODE STARTS HERE". Modifying code outside of these areas should be done with caution and with a deep understanding of how these changes may impact the EVM.
+Typically, custom codes are required in only those areas.
+4- Add your precompile upgrade in params/config.go
+5- Add your upgradable config in params/precompile_config.go
+6- Add your solidity interface and test contract to contract-examples/contracts
+7- Write solidity tests for your precompile in contract-examples/test
+8- Create e2e test for your solidity test in tests/e2e/solidity/suites.go
+9- Create your genesis with your precompile enabled in tests/e2e/genesis/
+10- Run your e2e precompile Solidity tests with 'E2E=true ./scripts/run.sh'
+
+*/
+
+package precompile
+
+import (
+	"math/big"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/ava-labs/subnet-evm/accounts/abi"
+	"github.com/ava-labs/subnet-evm/vmerrs"
+
+	"github.com/ethereum/go-ethereum/common"
+)
+
+const (
+	{{- range .Contract.Funcs}}
+	{{.Normalized.Name}}GasCost uint64 = 0 // SET A GAS COST HERE
+	{{- end}}
+	{{- if .Contract.Fallback}}
+	{{.Contract.Type}}FallbackGasCost uint64 = 0 // SET A GAS COST LESS THAN 2300 HERE
+  {{- end}}
+
+	// {{.Contract.Type}}RawABI contains the raw ABI of {{.Contract.Type}} contract.
+	{{.Contract.Type}}RawABI = "{{.Contract.InputABI}}"
+)
+
+// Reference imports to suppress errors if they are not otherwise used.
+var (
+	_ = errors.New
+	_ = big.NewInt
+	_ = strings.NewReader
+)
+
+{{$contract := .Contract}}
+// Singleton StatefulPrecompiledContract and signatures.
+var (
+	_ StatefulPrecompileConfig = &{{.Contract.Type}}Config{}
+
+	{{.Contract.Type}}Precompile StatefulPrecompiledContract = create{{.Contract.Type}}Precompile({{.Contract.Type}}Address)
```

Comment:
This needs to be moved into the `init` function since calling this here will take place before `init()` gets invoked.

This results in the ABI not being populated when the create function is called, such that each referenced `method.ID` is empty.

Reply:
To catch this in the future, we should also check that the method names actually exist within the `create` function so that if the requested function does not exist, it will detect it there immediately.

Reply:
fixed & added some checks

Reply:
Great, would you mind making a PR to geth as well with the moved error checks in the ABI package to fail earlier?

Reply:
yep i will also create one for coreth, or are we going to include this in precompile gen PR?